src/SourceBuild/content/test/Microsoft.DotNet.SourceBuild.SmokeTests/LicenseScanTests.cs

```diff
@@ -0,0 +1,243 @@
+    {
+        IEnumerable<string> rawExclusions = Utilities.ParseExclusionsFile("LicenseExclusions.txt");
+        IEnumerable<LicenseExclusion> exclusions = rawExclusions
+            .Select(exclusion => ParseLicenseExclusion(exclusion))
+            .Where(exclusion => exclusion.Repo == _targetRepo)
+            .ToList();
+
+        // This will filter out files that we don't want to include in the baseline.
+        // Filtering can happen in two ways:
+        //   1. There are a set of allowed license expressions that apply to all files. If a file has a match on one of those licenses,
+        //      that license will not be considered.
+        //   2. The LicenseExclusions.txt file contains a list of files and the licenses that should be excluded from those files.
+        // Once the license expression filtering has been applied, if a file has any licenses left, it will be included in the baseline.
+        // In that case, the baseline will list all of the licenses for that file, even if some were originally excluded during this processing.
+        // In other words, the baseline will be fully representative of the licenses that apply to the files that are listed there.
+
+        for (int i = scancodeResults.Files.Count - 1; i >= 0; i--)
+        {
+            ScancodeFileResult file = scancodeResults.Files[i];
+
+            // A license expression can be a logical expression, e.g. "(MIT OR Apache-2.0)"
+            // For our purposes, we just care about the license involved, not the semantics of the expression.
+            // Parse out all the expression syntax to just get the license names.
+            string[] licenses = file.LicenseExpression?
+                .Replace("(", string.Empty)
+                .Replace(")", string.Empty)
+                .Replace(" AND ", ",")
+                .Replace(" OR ", ",")
+                .Split(",", StringSplitOptions.RemoveEmptyEntries)
+                .Select(license => license.Trim())
+                .ToArray()
+                ?? Array.Empty<string>();
+
+            // First check whether the file's licenses can all be matched with allowed expressions
+            IEnumerable<string> disallowedLicenses = licenses
+                .Where(license => !s_allowedLicenseExpressions.Contains(license, StringComparer.OrdinalIgnoreCase));
+
+            if (!disallowedLicenses.Any())
+            {
+                scancodeResults.Files.Remove(file);
+            }
+            else
+            {
+                // There are some licenses that are not allowed. Now check whether the file is excluded.
+
+                IEnumerable<LicenseExclusion> matchingExclusions =
+                    Utilities.GetMatchingFileExclusions(file.Path, exclusions, exclusion => exclusion.Path);
+
+                if (matchingExclusions.Any())
+                {
+                    IEnumerable<string> excludedLicenses = matchingExclusions.SelectMany(exclusion => exclusion.LicenseExpressions);
```

Comment:
Can this be combined with matchingExclusions to simplify the code?  It would eliminate the need for `if (matchingExclusions.Any())` and remove a nesting level.

Reply:
Fixed in https://github.com/dotnet/installer/pull/17442/commits/0b07e41fde70a84c3bfc5f70720b4b7d5cf336bc